File: 0940-fruit-into-baskets/0940-fruit-into-baskets.js
/**
 * @param {number[]} fruits
 * @return {number}
 */
var totalFruit = function(fruits) {
    /*
        Sliding window
        - keep track of fruit types in window
        - if we reach more than 3 fruit types we shrink the window 
        - keep track of max window length
    */

    let maxFruit = 0;
    let L = 0;
    let types = new Map();
    
    for (let R = 0; R < fruits.length; R++) {
        types.set(fruits[R], (types.get(fruits[R]) || 0) + 1);
        while (types.size > 2) {
            types.set(fruits[L], types.get(fruits[L]) - 1);
            if (types.get(fruits[L]) === 0) types.delete(fruits[L]);
            L += 1;
        }
        maxFruit = Math.max(maxFruit, R - L + 1);
    }
    return maxFruit;
};

File: 0940-fruit-into-baskets/0940-fruit-into-baskets.py
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        L = 0
        maxFruit = 0
        types = {}

        for R in range(0, len(fruits)):
            types[fruits[R]] = types.get(fruits[R], 0) + 1
            while len(types) > 2:
                types[fruits[L]] = types.get(fruits[L], 0) - 1
                if (types.get(fruits[L]) == 0): types.pop(fruits[L], None)
                L += 1
            maxFruit = max(maxFruit, R - L + 1)
        
        return maxFruit